### PR TITLE
RATIS-1754. Set leader to null when stepping down

### PR DESCRIPTION
**Pull request by SzyWilliam**
## What changes were proposed in this pull request?
ServerState.getLeaderId() still returns itself when a leader just steps down. This PR proposes to update leader to null when stepping down.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/RATIS-1754

### CODE REVIEW

File: ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java [hunk 1]
```diff
@@ -568,6 +568,7 @@ void submitStepDownEvent(long term, StepDownReason reason) {
   private void stepDown(long term, StepDownReason reason) {
     try {
       server.changeToFollowerAndPersistMetadata(term, false, reason);
+      server.getState().setLeader(null, "stepDown");
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java:571]: Let's set it in `changeToFollower` as below
```java
+++ b/ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java
@@ -534,6 +534,7 @@ class RaftServerImpl implements RaftServer.Division,
       setRole(RaftPeerRole.FOLLOWER, reason);
       if (old == RaftPeerRole.LEADER) {
         role.shutdownLeaderState(false);
+        state.setLeader(null, reason);
       } else if (old == RaftPeerRole.CANDIDATE) {
         role.shutdownLeaderElection();
       } else if (old == RaftPeerRole.FOLLOWER) {
```

**Review by szetszwo — COMMENTED** [review]
@SzyWilliam , thanks a lot for working on this!  Let's also change `leaderId` to `AtomicReference<RaftPeerId>`; see https://issues.apache.org/jira/secure/attachment/13053249/792_review.patch

**Comment by SzyWilliam** [comment]
@szetszwo Thanks for the reviews! Applied the review patch.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Comment by szetszwo** [comment]
@SzyWilliam , since this set leaderId to null, it becomes easier to get NPE; see #794.

**Comment by szetszwo** [comment]
@SzyWilliam ,  `ReadOnlyRequestTests.testLinearizableReadTimeout` still fails with NPE, could you rebase this with the master branch?

**Pull request merged**